File: frontend/src/types/metric.ts
export interface Metric {
  id: string;
  userId: string;
  recordedBy: string;
  name: string;
  unit: string | null;
  value: number;
  recordedDate: string;
  higherIsBetter: boolean;
  source: string | null;
  comment: string | null;
}

export interface MetricInput {
  name: string;
  unit: string | null;
  value: number;
  recorded_date: string;
  higher_is_better: boolean;
  source: string | null;
  comment: string | null;
}

export interface PersonalRecord {
  name: string;
  unit: string | null;
  value: number;
  recordedDate: string;
  higherIsBetter: boolean;
}

export interface MetricDto {
  id: string;
  user_id: string;
  recorded_by: string;
  name: string;
  unit: string | null;
  value: number;
  recorded_date: string;
  higher_is_better: boolean;
  source: string | null;
  comment: string | null;
}

export interface PersonalRecordDto {
  name: string;
  unit: string | null;
  value: number;
  recorded_date: string;
  higher_is_better: boolean;
}

export function mapMetricDto(dto: MetricDto): Metric {
  return {
    id: dto.id,
    userId: dto.user_id,
    recordedBy: dto.recorded_by,
    name: dto.name,
    unit: dto.unit,
    value: dto.value,
    recordedDate: dto.recorded_date,
    higherIsBetter: dto.higher_is_better,
    source: dto.source,
    comment: dto.comment,
  };
}

export function mapPersonalRecordDto(dto: PersonalRecordDto): PersonalRecord {
  return {
    name: dto.name,
    unit: dto.unit,
    value: dto.value,
    recordedDate: dto.recorded_date,
    higherIsBetter: dto.higher_is_better,
  };
}

File: backend/app/schemas/metric.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class MetricIn(BaseModel):
    name: str = Field(min_length=1, max_length=150)
    unit: str | None = Field(default=None, max_length=30)
    value: float
    recorded_date: date
    higher_is_better: bool = True
    source: str | None = Field(default=None, max_length=150)
    comment: str | None = Field(default=None, max_length=1000)


class MetricOut(BaseModel):
    id: UUID
    user_id: UUID
    recorded_by: UUID
    name: str
    unit: str | None
    value: float
    recorded_date: date
    higher_is_better: bool
    source: str | None
    comment: str | None
    created_at: datetime
    updated_at: datetime


class PersonalRecordOut(BaseModel):
    name: str
    unit: str | None
    value: float
    recorded_date: date
    higher_is_better: bool
